src/js/tech/setup-sourceset.js

```diff
@@ -42,42 +161,175 @@ const setupSourceset = function(tech) {
+ * - `innerHTML` -  can be used to add a `<source>` element to the media element
+ *
+ * @param {Html5} tech
+ *        The tech object that sourceset is being setup on.
+ */
+const firstSourceWatch = function(tech) {
+  const el = tech.el();
+
+  if (el.firstSourceWatch_) {
+    return;
+  }
+
+  el.firstSourceWatch_ = true;
+  const oldAppend = el.append;
+  const oldAppendChild = el.appendChild;
+  const oldInsertAdjacentHTML = el.insertAdjacentHTML;
+  const innerDescriptor = getInnerHTMLDescriptor(el);
+
+  el.appendChild = function() {
+    const retval = oldAppendChild.apply(el, arguments);
+
+    sourcesetLoad(tech);
+
+    return retval;
+  };
+
+  if (oldAppend) {
+    el.append = function() {
+      const retval = oldAppend.apply(el, arguments);
+
+      sourcesetLoad(tech);
+
+      return retval;
+    };
+  }
+
+  if (oldInsertAdjacentHTML) {
+    el.insertAdjacentHTML = function() {
+      const retval = oldInsertAdjacentHTML.apply(el, arguments);
+
+      sourcesetLoad(tech);
+
+      return retval;
+    };
+  }
+
+  Object.defineProperty(el, 'innerHTML', {
+    get: innerDescriptor.get.bind(el),
+    set(v) {
+      const retval = innerDescriptor.set.call(el, v);
+
+      sourcesetLoad(tech);
+
+      return retval;
+    },
+    configurable: true,
+    enumerable: innerDescriptor.enumerable
+  });
+
+  // on the first sourceset, we need to revert
+  // our changes
+  tech.one('sourceset', (e) => {
+    el.firstSourceWatch_ = false;
+    el.appendChild = oldAppendChild;
+    if (oldAppend) {
+      el.append = oldAppend;
+    }
+    if (oldInsertAdjacentHTML) {
+      el.insertAdjacentHTML = oldInsertAdjacentHTML;
+    }
+
+    Object.defineProperty(el, 'innerHTML', {
+      get: innerDescriptor.get.bind(el),
+      set: innerDescriptor.set.bind(el),
+      configurable: true,
+      enumerable: innerDescriptor.enumerable
+    });
+  });
+};
+
+/**
+ * setup `sourceset` handling on the `Html5` tech. This function
+ * patches the following element properties/functions:
+ *
+ * - `src` - to determine when `src` is set
+ * - `setAttribute()` - to determine when `src` is set
+ * - `load()` - this re-triggers the source selection algorithm, and can
+ *              cause a sourceset.
+ *
+ * If there is no source when we are adding `sourceset` support or during a `load()`
+ * we also patch the functions listed in `firstSourceWatch`.
+ *
+ * @param {Html5} tech
+ *        The tech to patch
+ */
+const setupSourceset = function(tech) {
+  if (!tech.featuresSourceset) {
+    return;
+  }
+
+  const el = tech.el();
+  const srcDescriptor = getSrcDescriptor(el);
+  const oldSetAttribute = el.setAttribute;
+  const oldLoad = el.load;
+
+  // we need to fire sourceset when the player is ready
+  // if we find that the media element had a src when it was
+  // given to us and that tech element is not in a stalled state
+  if (el.src || el.currentSrc && el.initNetworkState_ !== 3) {
+    tech.triggerSourceset(el.src || el.currentSrc);
```

Comment:
should we just call `sourcesetLoad` here instead?

Reply:
yeah I think we should